arch/cortex-m4/src/mpu.rs

```diff
@@ -116,23 +118,55 @@ impl kernel::mpu::MPU for MPU {
         }
     }
 
-    fn set_mpu(&self,
-               region_num: u32,
-               start_addr: u32,
-               len: PowerOfTwo,
-               subregion_mask: u8,
-               execute: kernel::mpu::ExecutePermission,
-               access: kernel::mpu::AccessPermission) {
+    fn create_region(region_num: usize,
+                     start: usize,
+                     len: usize,
+                     execute: kernel::mpu::ExecutePermission,
+                     access: kernel::mpu::AccessPermission)
+                     -> Option<Region> {
+        if start % len != 0 {
+            // Text length not aligned to text start
+            let subregion_size = start % len;
+            let region_size = subregion_size * 8; // 8 subregions in a region
+            let region_start = start - (start % region_size);
+
+            if region_size + region_start - start < len {
+                return None;
+            }
+
+
+            let min_subregion = (start - region_start) / subregion_size;
+            let max_subregion = min_subregion + len / subregion_size - 1;
+
+            let region_len = PowerOfTwo::floor(region_size as u32);
+
+            let subregion_mask = (min_subregion..(max_subregion + 1))
+                .fold(!0, |res, i| res & !(1 << i)) & 0xff;
+            let xn = execute as u32;
+            let ap = access as u32;
+            Some(unsafe {
+                Region::new((region_start | 1 << 4 | (region_num & 0xf)) as u32,
+                            1 | subregion_mask << 8 | (region_len.exp::<u32>() - 1) << 1 |
+                            ap << 24 | xn << 28)
+            })
+
+        } else {
+            // Text length aligned to text start
+            let region_len = PowerOfTwo::floor(len as u32);
+            let xn = execute as u32;
+            let ap = access as u32;
+            Some(unsafe {
+                Region::new((start | 1 << 4 | (region_num & 0xf)) as u32,
+                            1 | (region_len.exp::<u32>() - 1) << 1 | ap << 24 | xn << 28)
```

Comment:
I was going to comment on the above `Region::new` that I felt like the region constructor function shouldn't be a simple `u32` of attributes, but Branden explained that the goal is for Region to be chip-agnostic.

In that case, I think it'd be valuable for this file to have a `fn new_region` or some such that breaks down all of the attributes into pieces. That way, you only have the bitshift logic in one place (thus far it's in at least two places) and also lets you consolidate the checks Branden suggested